iceoryx2-pal/posix/src/macos/mman.rs

```diff
@@ -74,23 +78,94 @@ unsafe fn shm_file_path(name: *const char, suffix: &[u8]) -> [u8; MAX_PATH_LENGT
     state_file_path
 }
 
+unsafe fn get_real_shm_name(name: *const char) -> Option<[u8; SHM_MAX_NAME_LEN]> {
+    let shm_file_path = shm_file_path(name, SHM_STATE_SUFFIX);
+    let shm_state_fd = open_with_mode(shm_file_path.as_ptr().cast(), O_RDONLY, 0);
+    if shm_state_fd == -1 {
+        return None;
+    }
+
+    let mut buffer = [0u8; SHM_MAX_NAME_LEN];
+
+    if read(shm_state_fd, buffer.as_mut_ptr().cast(), SHM_MAX_NAME_LEN) <= 0 {
+        close(shm_state_fd);
+        return None;
+    }
+
+    close(shm_state_fd);
+    Some(buffer)
+}
+
+unsafe fn write_real_shm_name(name: *const char, buffer: &[u8]) -> bool {
+    let shm_file_path = shm_file_path(name, SHM_STATE_SUFFIX);
+    let shm_state_fd = open_with_mode(
+        shm_file_path.as_ptr().cast(),
+        O_EXCL | O_CREAT | O_RDWR,
+        S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH,
+    );
+
+    if shm_state_fd == -1 {
+        return false;
+    }
+
+    if write(shm_state_fd, buffer.as_ptr().cast(), buffer.len()) != buffer.len() as _ {
+        remove(shm_file_path.as_ptr().cast());
+        close(shm_state_fd);
+        return false;
+    }
+
+    close(shm_state_fd);
+    true
+}
+
+unsafe fn generate_real_shm_name() -> [u8; SHM_MAX_NAME_LEN] {
+    let mut now = timespec::new();
+    clock_gettime(CLOCK_REALTIME, &mut now);
+    let pid = getpid();
+    let shm_name = pid.to_string() + "_" + &now.tv_sec.to_string() + "_" + &now.tv_nsec.to_string();
+
+    let mut buffer = [0u8; SHM_MAX_NAME_LEN];
+    buffer.as_mut_slice().write_all(shm_name.as_bytes()).expect(
+        "always works since 3 32-bit numbers and 2 underscores do not use more than 33 characters",
+    );
+    buffer
+}
+
 pub unsafe fn shm_open(name: *const char, oflag: int, mode: mode_t) -> int {
-    let ret_val = crate::internal::shm_open(name, oflag, mode as uint);
-    if ret_val != -1 {
-        let shm_file_path = shm_file_path(name, SHM_STATE_SUFFIX);
+    let real_name = get_real_shm_name(name);
+    if oflag & O_EXCL != 0 && real_name.is_some() {
+        Errno::set(Errno::EEXIST);
+        return -1;
+    }
 
-        if open_with_mode(shm_file_path.as_ptr().cast(), O_CREAT | O_RDONLY, mode) == -1 {
-            close(ret_val);
-            return -1;
+    let real_name = match real_name {
+        Some(name) => name,
+        None => {
+            if oflag & O_CREAT == 0 {
+                Errno::set(Errno::ENOENT);
+                return -1;
+            }
+            let real_name = generate_real_shm_name();
+            if !write_real_shm_name(name, &real_name) {
+                return -1;
+            }
+            real_name
         }
-    }
+    };
 
-    ret_val
+    crate::internal::shm_open(real_name.as_ptr().cast(), oflag, mode as uint)
 }
 
 pub unsafe fn shm_unlink(name: *const char) -> int {
+    let real_name = get_real_shm_name(name);
     remove(shm_file_path(name, SHM_STATE_SUFFIX).as_ptr().cast());
```

Comment:
Wouldn't it be better to delete this file only after the shm is closed. Just in case there is an error on closing the shm and one needs to retry?

Reply:
No, since a shared memory is persistent and it is useful to keep this file until the shared memory is completely removed otherwise we may leak shared memory since we no longer know their names.
There is also no possibility in mac os to list all shared memories.

Reply:
Maybe I got this wrong. I assumed that this line deletes the file with the real shm name.

Reply:
No it deletes the real shm name. `shm_unlink` removes the shared memory so the file must be deleted. If you have still opened the shm you are not affected since the OS does an ref count with the fds.

Reply:
Maybe I still get it wrong. My concern is that this file gets deleted and then `shm_unlink` does not work and the shm is left in the filesystem without a possibility to remove it.

Reply:
You are completely right, this would be a problem. I adjusted the behavior, now the state file is only removed when either `shm_unlink` was successful or it failed with the errno `ENOENT` signaling that it does not exist.